core/upgrader/impl/src/services/disaster_recovery.rs

```diff
@@ -251,63 +254,75 @@ impl DisasterRecoveryService {
 
             if since + DISASTER_RECOVERY_IN_PROGESS_EXPIRATION_NS > time() {
                 logger.log(LogEntryType::DisasterRecoveryInProgress(log));
-                return;
+                return Ok(());
             }
 
             logger.log(LogEntryType::DisasterRecoveryInProgressExpired(log));
             value.recovery_status = RecoveryStatus::Idle;
         }
 
-        let Some(station_canister_id) =
-            TARGET_CANISTER_ID.with(|id| id.borrow().get(&()).map(|id| id.0))
-        else {
-            value.last_recovery_result = Some(RecoveryResult::Failure(RecoveryFailure {
-                reason: "Station canister ID not set".to_string(),
-            }));
-            storage.set(value);
-            return;
-        };
-
         value.recovery_status = RecoveryStatus::InProgress { since: time() };
         storage.set(value);
 
+        let station_canister_id = TARGET_CANISTER_ID
+            .with(|id| id.borrow().get(&()).map(|id| id.0))
+            .ok_or("Station canister ID not set")?;
+
         let mut releaser = DisasterRecoveryReleaser {
             storage: storage.clone(),
             result: None,
             logger: logger.clone(),
         };
 
-        // only stop for upgrade
-        if request.install_mode == InstallMode::Upgrade {
-            if let Err(err) = installer.stop(station_canister_id).await {
-                ic_cdk::print(err);
+        if let Err(err) = installer.stop(station_canister_id).await {
+            if request.force {
+                let existing_snapshots = list_canister_snapshots(CanisterIdRecord {
+                    canister_id: station_canister_id,
+                })
+                .await
+                .map_err(|(_code, msg)| msg)?
+                .0;
+                let snapshot = take_canister_snapshot(TakeCanisterSnapshotArgs {
+                    canister_id: station_canister_id,
+                    replace_snapshot: existing_snapshots
+                        .into_iter()
+                        .next()
+                        .map(|snapshot| snapshot.id),
+                })
+                .await
+                .map_err(|(_code, msg)| msg)?
+                .0;
+                uninstall_code(CanisterIdRecord {
+                    canister_id: station_canister_id,
+                })
+                .await
+                .map_err(|(_code, msg)| msg)?;
+                load_canister_snapshot(LoadCanisterSnapshotArgs {
+                    canister_id: station_canister_id,
+                    snapshot_id: snapshot.id,
+                    sender_canister_version: None,
+                })
+                .await
+                .map_err(|(_code, msg)| msg)?;
```

Comment:
if this would fail for some reason (e.g. because maybe the target canister would not have enough cycles), then the disaster recovery committee would need to call this endpoint again, however, because we replace that snapshot, then the second try would take a snapshot of an empty canister because we would have already uninstalled it.

To account for such cases, maybe we should store a flag in the upgrader that makes it reuse the same snapshot id on a retry, wdyt?

Reply:
Great catch! I'd suggest the following:
- support taking and restoring snapshots manually in disaster recovery
- never take a snapshot automatically: in the force mode a manually created snapshot must be specified, i.e., `force: bool` becomes `force: Option<SnapshotId>`

Reply:
Both look like a safer approach, and in the case of `force: Option<SnapshotId>`, if provided as `None` we would then create the snapshot? as it is the behaviour for `force=true`

Reply:
`force: None` would correspond to `force: false` and `force: Some(snapshot_id)` would correspond to `force: true`